telegram/telegram.py

```diff
@@ -0,0 +1,194 @@
+import json
+
+from treq.client import HTTPClient
+
+from twisted.internet import reactor
+from twisted.internet.defer import inlineCallbacks
+from twisted.web import http
+from twisted.web.client import Agent
+from twisted.web._newclient import ResponseFailed
+
+from vumi.transports.httprpc.httprpc import HttpRpcTransport
+from vumi.config import ConfigText
+from vumi import log
+
+
+class TelegramTransportConfig(HttpRpcTransport.CONFIG_CLASS):
+    bot_username = ConfigText(
+        'The username of our Telegram bot',
+        static=True,
+        required=True,
+    )
+    bot_token = ConfigText(
+        "Our bot's unique token to access the Telegram API",
+        static=True,
+        required=True,
+    )
+    outbound_url = ConfigText(
+        'The URL our bot should make requests to',
+        default='https://api.telegram.org/bot', static=True,
+    )
+    inbound_url = ConfigText(
+        'The URL our transport will listen on for Telegram updates',
+        static=True, required=True,
+    )
+
+
+class TelegramTransport(HttpRpcTransport):
+    """
+    Telegram transport for Vumi
+    """
+    transport_type = 'telegram'
+    transport_name = 'telegram_transport'
+
+    CONFIG_CLASS = TelegramTransportConfig
+
+    @classmethod
+    def agent_factory(cls):
+        """For swapping out the Agent for use in tests"""
+        return Agent(reactor)
+
+    @inlineCallbacks
+    def setup_webhook(self):
+        # NOTE: Telegram currently only supports ports 80, 88, 443 and 8443 for
+        #       webhook setup
+        url = '%s/setWebhook' % self.outbound_url.rstrip('/')
+        http_client = HTTPClient(self.agent_factory())
+
+        try:
+            r = yield http_client.post(
+                url=url,
+                data=json.dumps({'url': self.inbound_url}),
+                headers={'Content-Type': ['application/json']}
+                )
+            content = yield r.content()
+            res = json.loads(content)
+            if r.code == http.OK and res['ok']:
+                log.info('Webhook set up on %s' % self.inbound_url)
+            else:
+                log.info('Webhook setup failed: %s' % res['description'])
+
+        # Treat page redirects as errors, since Telegram seems to redirect us
+        # when our bot token is invalid
+        except ResponseFailed as e:
+            log.info(
```

Comment:
These `log.info` should be `self.log.info` -- the base Vumi worker sets up a special `WrappingLogger` that sets the `system_name` of the logger to the transport name. This is used by Junebug to catch logs written by the transport.